app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,237 @@
+        this.serviceHandler = NetworkRequestHandler(context, handlerThread.looper, this)
+    }
+
+    interface NetworkListener {
+        fun onNetworkDisconnected()
+        fun onNetworkConnected(networks: LinkedHashSet<Network>?)
+    }
+
+    fun removeCallBack() {
+        connectivityManager.unregisterNetworkCallback(this)
+        destroy()
+    }
+
+    private fun destroy() {
+        this.serviceHandler?.removeCallbacksAndMessages(null)
+        this.handlerThread.quitSafely()
+        this.serviceHandler = null
+    }
+
+    override fun onAvailable(network: Network) {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - onAvailable")
+        handleNetworkChange()
+    }
+
+    override fun onLost(network: Network) {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - onLost")
+        handleNetworkChange()
+    }
+
+    override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - onCapabilitiesChanged")
+        handleNetworkChange()
+    }
+
+    override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - onLinkPropertiesChanged")
+        handleNetworkChange()
+    }
+
+    /**
+     * Handles user preference changes, ie, when the user elects to see
+     * either multiple underlying networks, or just one (the active network).
+     */
+    fun onUserPreferenceChanged() {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - onUserPreferenceChanged")
+        handleNetworkChange(isForceUpdate = true)
+    }
+
+    /**
+     * Force updates the VPN's underlying network based on the preference.
+     * Will be initiated when the VPN start is completed.
+     */
+    fun onVpnStarted() {
+        if (DEBUG) Log.d(LOG_TAG, "$FILE_LOG_TAG - new vpn is created force update the network")
+        handleNetworkChange(isForceUpdate = true)
+    }
+
+    private fun handleNetworkChange(isForceUpdate: Boolean = false) {
+        val message = constructMessage(if (persistentState.isAddAllNetworks) MSG_ADD_ALL_NETWORKS else MSG_ADD_ACTIVE_NETWORK, isForceUpdate)
+        serviceHandler?.removeMessages(MSG_ADD_ACTIVE_NETWORK, null)
+        serviceHandler?.removeMessages(MSG_ADD_ALL_NETWORKS, null)
+        serviceHandler?.sendMessage(message)
+    }
+
+    /**
+     * Constructs the message object for Network handler.
+     * Add the active network to the message object
+     * in case of setUnderlying network has only active networks.
+     */
+    private fun constructMessage(what: Int, isForceUpdate: Boolean): Message {
+        val message = Message.obtain()
+        message.what = what
+        message.obj = isForceUpdate
+        return message
+    }
+
+    // Handles the network messages from the callback from the connectivity manager
+    private class NetworkRequestHandler(context: Context, looper: Looper, val connectionMonitor: ConnectionMonitor) : Handler(looper) {
```

Comment:
Don't really need to depend on `ConnectionMonitor` just the `NetworkListener`?

Reply:
Modified.